app/explore/_features/browse-projects-filters/browse-projects-filters.context.tsx

```diff
@@ -0,0 +1,86 @@
+import { createContext, useContext, useEffect, useMemo, useState } from "react";
+import { useDebounce } from "react-use";
+
+import {
+  BrowseProjectsContextFilter,
+  BrowseProjectsContextProviderProps,
+  BrowseProjectsContextQueryParams,
+  BrowseProjectsContextReturn,
+  DEFAULT_FILTER,
+} from "./browse-projects-filters.types";
+
+const BrowseProjectsContext = createContext<BrowseProjectsContextReturn>({
+  filters: {
+    values: DEFAULT_FILTER,
+    isCleared: true,
+    set: () => null,
+    clear: () => null,
+    count: 0,
+  },
+  queryParams: {},
+});
+
+export function BrowseProjectsContextProvider({ children }: BrowseProjectsContextProviderProps) {
+  const [filters, setFilters] = useState<BrowseProjectsContextFilter>(DEFAULT_FILTER);
+  const [queryParams, setQueryParams] = useState<BrowseProjectsContextQueryParams>({});
+  const [debouncedQueryParams, setDebouncedQueryParams] = useState<BrowseProjectsContextQueryParams>(queryParams);
+
+  useDebounce(
+    () => {
+      setDebouncedQueryParams(queryParams);
+    },
+    300,
+    [queryParams]
+  );
+
+  useEffect(() => {
+    setQueryParams({
+      tags: filters.tags.length ? filters.tags : undefined,
+      languageSlugs: filters.languageSlugs.length ? filters.languageSlugs : undefined,
+      ecosystemSlugs: filters.ecosystemSlugs.length ? filters.ecosystemSlugs : undefined,
+      categorySlugs: filters.categorySlugs.length ? filters.categorySlugs : undefined,
+    });
```

Comment:
this should go inside the setFilter handler instead of use a async effect

Reply:
Then I would have to do this
```
  function setFilter(filter: Partial<BrowseProjectsContextFilter>) {
    const newFilters = { ...filters, ...filter };

    setFilters(newFilters);

    setQueryParams({
      tags: newFilters.tags.length ? newFilters.tags : undefined,
      languageSlugs: newFilters.languageSlugs.length ? newFilters.languageSlugs : undefined,
      ecosystemSlugs: newFilters.ecosystemSlugs.length ? newFilters.ecosystemSlugs : undefined,
      categorySlugs: newFilters.categorySlugs.length ? newFilters.categorySlugs : undefined,
    });
  }

  function clearFilters() {
    setFilters(DEFAULT_FILTER);

	setQueryParams({
		tags: DEFAULT_FILTER.tags.length ? DEFAULT_FILTER.tags : undefined,
		languageSlugs: DEFAULT_FILTER.languageSlugs.length ? DEFAULT_FILTER.languageSlugs : undefined,
		ecosystemSlugs: DEFAULT_FILTER.ecosystemSlugs.length ? DEFAULT_FILTER.ecosystemSlugs : undefined,
		categorySlugs: DEFAULT_FILTER.categorySlugs.length ? DEFAULT_FILTER.categorySlugs : undefined,
	  });
  }
```